### ISSUE TITLE
[Feature Request] Helper function assert

### ISSUE THREAD

**Issue by Xerkus** [hashicorp/consul-template#1542]
I would like to request an `assert` helper function. I think consul-template lacks the ability to conditionally fail rendering when result would otherwise be unacceptable. 

TLDR:

```
{{ assert "message" }}
{{ assert "message" false }}

{{ assert "Must be at least 3" (ge . 3) }}
{{ regexMatch "[a-z]+" | assert "Must match regex" }}
{{ with . | assert "Pipe pass-through assert" ( contains "production" . ) | toJson }}
```

More practical example is using template stanza in nomad task definition to render job configuration:

```hcl
template {
  data = <<EOH
  {{ with service "rabbtimq" }}{{ with index . 0 }}
  RABBITMQ_ADDR = "{{ .Address }}"
  RABBITMQ_PORT = "{{ .Port }}"
  {{ end }}{{ end }}
  EOH

  destination = "secrets/queue.env"
  env = true
}
```

This template will succeed rendering but it will be missing important task configuration. With no RabbitMQ available there is no point starting the worker. What is worse, the template example I provided actually have typo in service name, which is hard to spot, so there would never be a result but task will start anyway only to fail .

Ability to conditionally fail template rendering with a message would be very useful:

```hcl
template {
  data = <<EOH
  {{ with service "rabbtimq" }}{{ with index . 0 }}
  RABBITMQ_ADDR = "{{.Address}}"
  RABBITMQ_PORT = "{{.Port}}"
  {{end}}
  {{ else }}{{ assert "Missing rabbitmq service" }}
  {{end}}
  EOH

  destination = "secrets/queue.env"
  env = true
}
```

**Comment by Xerkus** [comment]
I just realized that last example won't work - it will always cause failure on first pass.

**Comment by eikenb** [comment]
Hey @Xerkus, thanks for the submission.

Would you consider updating your original post with changes reflecting your realization? It'd be much nicer to evaluate knowing it is all correct/what you want. Thanks!

**Comment by Xerkus** [comment]
I do not understand how it works entirely.
My current understanding is that rendering happens twice - without data and on second pass with actual data.
That means that first pass will always visit `{{else}}` and trigger assert, thus aborting execution.

Disabling assert failures until second pass could be an option, I suppose?
Then it could actually work.

**Comment by eikenb** [comment]
You are correct that it wouldn't work and about the basics of rendering. CT doesn't know anything about what values to get until it runs once. It uses the initial run to find out about the consul fields it needs to monitor and start up those processes. Once values come in it will re-render the template. Depending on how many fields monitored and how they are organized, it might evaluate some values as empty more than once if they take more time to return their values.

Go 1.18 should be tweaking templates `and` and `or` functions so that they will short-circuit evaluation (stop evaluating when an `and` is false or an `or` is true) and might help facilitate this sort of functionality without adding a new keyword.
